Fix/registration

mspiridonov2706 opened the pull request:
Задача:
https://www.notion.so/Bot-a96778a3d4354e6c84e0561bdbbec551

1) Добавил вызов функции register из хэндлера **start**, что бы пользователю сразу предлагалось зарегистрироваться. До этого функция никак не вызывалась и под неё отсутствуют хэндлеры.
2) Добавил у бота новый хэндлер для взаимодействия с веб данными `bot_instance.add_handler(MessageHandler(StatusUpdate.WEB_APP_DATA, web_app_data))`. Вызывается при заполнении формы.
3) Добавил `app.mount("/static", StaticFiles(directory="src/html"), name="html template")` при создании приложения что бы бот мог получить доступ к файлу шаблона при регистрации
3) В функции register изменил путь к шаблону регистрации.

Фиксы ошибок во время валидации формы:
1) Добавил валидацию поля date_of_birth в модели  **UserCreateRequest** для приведения строки даты из формы к типу date. Иначе вылезает ошибка валидации:

**Ошибка при валидации данных: 1 validation error for UserCreateRequest
date_of_birth
  invalid date format (type=value_error.date)**

Причём именно в таком виде ошибка выводилась пользователю.

2) Добавил в блок try/except у функции register обработку исключения ValueError. Иначе сообщение о том что пользователь должен быть старше 3-х лет не выводится пользователю и после отправки формы он не видел никакие другие сообщения.

3) Изменил в функции check_user_existence запрос в базу данных, иначе срабатывает исключение во время запроса.

Commit:
fix: fix registration

Add static mount to get access to html template;
Add date_of_birth validator to UserCreateRequest Model for castsing date from form to python date type;
Add call register func from start;
Change template url in register func;
Add new exception in try except statement in web_app_data func;
Add new handler web_app_data in create bot func;
Change query in check_user_existence;
Small fixes.

Commit:
Merge branch 'develop' into fix/registration

Commit:
fix: fix codestyle

Commit:
fix: check_user_existence query

Commit:
fix: fix codestyle

ADesBiysk commented on src/core/db/repository/user_repository.py:
```diff
@@ -21,7 +21,7 @@ async def get_by_telegram_id(self, telegram_id: int) -> Optional[User]:
     async def check_user_existence(self, telegram_id: int, phone_number: str) -> bool:
         user_exists = await self._session.execute(
-            exists().where(or_(User.phone_number == phone_number, User.telegram_id == telegram_id))
+            exists().where(or_(User.phone_number == phone_number, User.telegram_id == telegram_id)).select()
```
Судя по документации должен был работать. 
Тогда, мне кажется, такой вариант логичнее:
select().where(or_(User.phone_number == phone_number, User.telegram_id == telegram_id)).exists()

mspiridonov2706 replied:
Опять появляется ошибка как и при первом варианте.

`Executable SQL or text() construct expected, got <sqlalchemy.sql.selectable.Exists object at 0x111423640>.`

mspiridonov2706 replied:
Тогда предлагаю такой вариант:
`select(select(User).where(or_(User.phone_number == phone_number, User.telegram_id == telegram_id)).exists())`
Что по сути является тем же самым что и текущий вариант. Либо использоваться flexibility вариант из документации 

https://docs.sqlalchemy.org/en/14/core/selectable.html#sqlalchemy.sql.expression.exists

`exists_criteria = (select(table2.c.col2).where(table1.c.col1 == table2.c.col2).exists())`
`stmt = select(table1.c.col1).where(exists_criteria)`

ADesBiysk commented on src/application.py:
```diff
@@ -21,6 +22,8 @@ def create_app() -> FastAPI:
 
     origins = ["*"]
 
+    app.mount("/static", StaticFiles(directory="src/html"), name="html template")
```
static лучше вынести в settings со значением по-умолчанию.

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -49,7 +50,7 @@ async def register(update: Update, context: CallbackContext) -> None:
             KeyboardButton(
                 text="Зарегистрироваться в проекте",
-                web_app=WebAppInfo(url=settings.APPLICATION_URL + "/lomaya_baryery_backend/src/html/registration.html"),
+                web_app=WebAppInfo(url=settings.APPLICATION_URL + "/static/registration.html"),
```
Лучше использовать f-string.

ADesBiysk commented on src/bot/handlers.py:
```diff
@@ -67,7 +68,7 @@ async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> No
         )
-    except ValidationError as e:
+    except (ValidationError, ValueError) as e:
         await update.message.reply_text(f"Ошибка при валидации данных: {e}")
```
Согласен. Странное решение. Лучше пользователю возвращать "человеческое" сообщение об ошибке вытащив имя поля из e.json().

Commit:
fix: review note

Move static data to settings;
Add f-string in url on WebAppInfo;
Fix errors messages for user in registration proccess;
Change check_user_existence query;

mspiridonov2706 commented on src/bot/handlers.py:
```diff
@@ -69,7 +69,9 @@ async def web_app_data(update: Update, context: ContextTypes.DEFAULT_TYPE) -> No
-        await update.message.reply_text(f"Ошибка при валидации данных: {e}")
+        if isinstance(e, ValidationError):
+            e = "\n".join([error['msg'] for error in json.loads(e.json())])
+        await update.message.reply_text(f"Ошибка при заполнении данных:\n{e}")
```
<img width="500" alt="Screen Shot 2022-11-19 at 00 28 17" src="https://user-images.githubusercontent.com/70363820/202809798-cd6ff3f6-ef5f-403f-866c-48be8dc8c9a7.png">
<img width="289" alt="Screen Shot 2022-11-19 at 00 50 03" src="https://user-images.githubusercontent.com/70363820/202809801-a5dc2c3a-6dab-419f-a208-eceb83a31258.png">


Теперь сообщения об ошибках выглядят так

FinemechanicPub replied:
json.loads(e.json()) выглядит странно. Если посмотреть исходный код, видно, что e.json() сбрасывает в строку словарь, получаемый от e.errors(). Может быть можно напрямую получать словарь из e.errors()?

B ещё один вопрос. Ведь такое преобразование сообщений об ошибках придется делать везде. В библиотеке бота не предусмотрено какой-то централизованной точки обработки исключений?

mspiridonov2706 replied:
Переделаю, просто до этого попросили брать e.json() и я как-то на нём зациклился.

Пока это вроде единственное место с обработкой исключений у бота при взаимодействии с пользователем.

ADesBiysk replied:
Ошибки могут возникнуть еще во время загрузки фотоотчёта. В любом случае сделать единственную точку обработки ошибок имеет смысл, но делать это будем позже. Сейчас надо довести проект до минимально рабочего состояния.

Commit:
fix: change list to tuple

NataliyaPavlova commented on src/bot/handlers.py:
```diff
@@ -49,7 +50,7 @@ async def register(update: Update, context: CallbackContext) -> None:
             KeyboardButton(
                 text="Зарегистрироваться в проекте",
-                web_app=WebAppInfo(url=settings.APPLICATION_URL + "/lomaya_baryery_backend/src/html/registration.html"),
+                web_app=WebAppInfo(url=f"{settings.APPLICATION_URL}/static/registration.html"),
```
вынеси всю ссылку в settings  сделай так же, как там для BOT_PERSISTENCE_FILE

NataliyaPavlova commented on src/core/services/user_service.py:
```diff
@@ -41,7 +41,7 @@ def __init__(
         """Регистрация пользователя. Отправка запроса на участие в смене."""
         user_scheme = UserCreateRequest(**user_data)
-        await validate_user_create(user_scheme, self.user_repository)
+        await validate_user_create(user_scheme, self.__user_repository)
```
за дебаг тебе огромное спасибо и плюс в карму

NataliyaPavlova commented on src/application.py:
```diff
@@ -21,6 +22,8 @@ def create_app() -> FastAPI:
 
     origins = ["*"]
 
+    app.mount(settings.STATIC_URL, StaticFiles(directory=settings.STATIC_PATH), name="static")
```
а надо же тогда еще подредактировать deploy workfile? создать диру static и скопировать туда html?

mspiridonov2706 replied:
ну по идее и без этого должно работать, я ведь указываю путь до папки где уже лежит html файл (src/html). А она уже и так копируется в контейнер. Или нет?

NataliyaPavlova replied:
он есть в контейнере, но нет на сервере.

Commit:
fix: review note 2

Fix WebAppInfo url;
Change e.json() on e.errors();
